tests/integration/relations/opensearch_provider/application-charm/src/charm.py

```diff
@@ -0,0 +1,216 @@
+
+    def connection_check(self) -> bool:
+        """Simple connection check to see if backend exists and we can connect to it."""
+        relations = self.model.relations.get("first-database", []) + self.model.relations.get(
+            "second-database", []
+        )
+        if len(relations) == 0:
+            return False
+        for relation in relations:
+            if not self.smoke_check(relation.id):
+                return False
+        return True
+
+    def smoke_check(self, relation_id) -> bool:
+        try:
+            resp = self.relation_request(relation_id, "GET", "/")
+            logger.info(resp)
+            return bool(resp)
+        except (OpenSearchHttpError, Exception) as e:
+            logger.error(e)
+            return False
+
+    # First database events observers.
+    def _on_first_database_created(self, event: DatabaseCreatedEvent) -> None:
+        """Event triggered when a database was created for this application."""
+        logging.info(f"first database credentials: {event.username} {event.password}")
+
+    def _on_first_database_endpoints_changed(self, event: DatabaseEndpointsChangedEvent) -> None:
+        """Event triggered when the read/write endpoints of the database change."""
+        logger.info(f"first database endpoints have been changed to: {event.endpoints}")
+
+    # Second database events observers.
+    def _on_second_database_created(self, event: DatabaseCreatedEvent) -> None:
+        """Event triggered when a database was created for this application."""
+        logger.info(f"second database credentials: {event.username} {event.password}")
+
+    def _on_second_database_endpoints_changed(self, event: DatabaseEndpointsChangedEvent) -> None:
+        """Event triggered when the read/write endpoints of the database change."""
+        logger.info(f"second database endpoints have been changed to: {event.endpoints}")
+
+    def _on_run_query_action(self, event: ActionEvent):
+        """Runs queries."""
+        raise NotImplementedError
+
+    def relation_request(
+        self,
+        relation_id: int,
+        method: str,
+        endpoint: str,
+        payload: Optional[Dict[str, any]] = None,
+    ) -> Union[Dict[str, any], List[any]]:
+        """Make an HTTP request to a specific relation."""
+        databag = self.first_database.fetch_relation_data()[relation_id]
+        logging.error(databag)
+        username = databag.get("username")
+        password = databag.get("password")
+        endpoints = databag.get("endpoints", "").split(",")
+
+        if None in [username, password] or len(endpoints) == 0:
+            raise OpenSearchHttpError
+
+        host = endpoints[0].split(":")[0]
+        port = int(endpoints[0].split(":")[1])
+
+        return self.request(
+            method,
+            endpoint,
+            port,
+            username,
+            password,
+            host,
+            payload=payload,
+        )
+
+    def request(
```

Comment:
Not urgent - But given this is a client app, it should be safe to remove this method and make use of one of the opensearch higher level python clients.

Reply:
good idea - this is just a basic implementation, but I'll switch it to a different client in the testing PR